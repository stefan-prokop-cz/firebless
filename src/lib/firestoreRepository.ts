import CollectionReference = FirebaseFirestore.CollectionReference;
import SetOptions = FirebaseFirestore.SetOptions;
import Query = FirebaseFirestore.Query;
import Precondition = FirebaseFirestore.Precondition;
import {
    bulkCreate,
    bulkDelete,
    bulkUpdate,
    create,
    detail,
    detailById,
    getModelData,
    list,
    Pagination,
    remove,
    removeById,
    update,
    updateById,
    upsert,
} from './queries';

export const bind = <Model extends { id?: string }>(collection: CollectionReference) => {
    return {
        bulkCreate: (data?: Partial<Model>[]) => bulkCreate<Model>(collection, data),
        bulkUpdate: (data?: Partial<Model>[], options?: SetOptions) => bulkUpdate<Model>(collection, data, options),
        bulkDelete: (ids?: string[], precondition?: Precondition) => bulkDelete(collection, ids, precondition),
        create: async (data?: Partial<Model>) => {
            const doc = await create<Model>(collection, data);
            return getModelData(doc) as Model;
        },
        delete: (params?: Partial<Model>, options?: Query) => remove<Model>(collection, params, options),
        deleteById: (id: string, precondition?: Precondition) => removeById(collection, id, precondition),
        detail: async (params?: Partial<Model>, options?: Query) => {
            const doc = await detail<Model>(collection, params, options);
            return getModelData<Model>(doc);
        },
        detailById: async (id: string) => {
            const doc = await detailById<Model>(collection, id);
            return getModelData<Model>(doc);
        },
        getCollection: () => collection,
        getFirestore: () => collection.firestore,
        list: async (filters?: Partial<Model> & Pagination, options?: Query) => {
            const docs = await list<Model>(collection, filters, options);
            return docs.map(doc => getModelData<Model>(doc)) as Model[];
        },
        update: (params?: Partial<Model>, data?: Partial<Model>, options?: SetOptions) => update<Model>(collection, params, data, options),
        updateById: async (id: string, data?: Partial<Model>, options?: SetOptions) => {
            const doc = await updateById<Model>(collection, id, data, options);
            return getModelData(doc) as Model;
        },
        upsert: async (data?: Partial<Model>, options?: SetOptions) => {
            const doc = await upsert<Model>(collection, data, options);
            return getModelData(doc) as Model;
        },
    };
};
