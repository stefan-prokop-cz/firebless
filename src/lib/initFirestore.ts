import { Firestore, Settings } from '@google-cloud/firestore';

export default (config: Settings) => new Firestore(config);
