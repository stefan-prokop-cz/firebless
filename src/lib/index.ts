import * as defaultFirestoreRepository from './firestoreRepository';
export { default as initFirestore } from './initFirestore';
export {
    applyFilters,
    applyPagination,
    bulkCreate,
    bulkDelete,
    bulkUpdate,
    create,
    detail,
    detailById,
    getDocument,
    getModelData,
    list,
    Pagination,
    remove,
    removeById,
    update,
    updateById,
    upsert,
} from './queries';
export { defaultFirestoreRepository };
