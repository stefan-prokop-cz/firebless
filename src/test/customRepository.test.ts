// @ts-ignore
import MockFirebase from 'mock-cloud-firestore';
import { defaultFirestoreRepository } from '../lib/index';

let fixtures = {
    __collection__: {
        users: {
            __doc__: {
                user_a: {
                    age: 15,
                    name: 'user',
                    surname: 'a',
                    birthDate: new Date('2019-01-01'),
                },
                user_b: {
                    age: 30,
                    name: 'user',
                    surname: 'a',
                    birthDate: new Date('2018-01-01'),
                },
            },
        },
    },
};

const firebase = new MockFirebase(fixtures);
const firestore = firebase.firestore();

interface User {
    id?: string;
    name: string;
    surname: string;
    age: number;
    birthDate: Date;
}

const userRepository = (defaultRepo => {
    const list = () => {
        const query = defaultRepo.getCollection().where('age', '>', 15);
        return defaultRepo.list({ surname: 'a' }, query);
    };
    return {
        ...defaultRepo,
        list,
    }
})(defaultFirestoreRepository.bind<User>(firestore.collection('users')));

describe('Custom repository', () => {
    it('List', async () => {
        const users = await userRepository.list();
        expect(Array.isArray(users)).toBe(true);
        expect(users).toHaveLength(1);
        expect(users[0]).toMatchSnapshot();
    });
    it('List 1', async () => {
        await userRepository.create({ age: 25, birthDate: new Date('2020-01-01'), name: 'name', surname: 'a' });
        const users = await userRepository.list();
        expect(Array.isArray(users)).toBe(true);
        expect(users).toHaveLength(2);
    });
});
