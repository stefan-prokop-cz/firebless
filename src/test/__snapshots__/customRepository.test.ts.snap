// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Custom repository List 1`] = `
Object {
  "age": 30,
  "birthDate": Object {
    "toDate": [Function],
  },
  "createdAt": undefined,
  "id": "user_b",
  "name": "user",
  "surname": "a",
  "updatedAt": undefined,
}
`;
